core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala

```diff
@@ -98,55 +99,92 @@ object TransactionMarkerChannelManager {
 
 class TransactionMarkerChannelManager(config: KafkaConfig,
                                       metadataCache: MetadataCache,
-                                      txnMarkerPurgatory: DelayedOperationPurgatory[DelayedTxnMarker],
-                                      interBrokerSendThread: InterBrokerSendThread,
-                                      transactionMarkerChannel: TransactionMarkerChannel) extends Logging {
-
-  type WriteTxnMarkerCallback = Errors => Unit
+                                      txnStateManager: TransactionStateManager,
+                                      txnMarkerSendThread: InterBrokerSendThread,
+                                      txnMarkerChannel: TransactionMarkerChannel,
+                                      txnMarkerPurgatory: DelayedOperationPurgatory[DelayedTxnMarker]) extends Logging {
 
   def start(): Unit = {
-    interBrokerSendThread.start()
+    txnMarkerSendThread.start()
   }
 
   def shutdown(): Unit = {
-    interBrokerSendThread.shutdown()
-    transactionMarkerChannel.clear()
+    txnMarkerSendThread.shutdown()
+    txnMarkerChannel.clear()
   }
 
-
-  def addTxnMarkerRequest(txnTopicPartition: Int, metadata: TransactionMetadata, coordinatorEpoch: Int, completionCallback: WriteTxnMarkerCallback): Unit = {
-    val metadataToWrite = metadata synchronized metadata.copy()
-
-    if (!transactionMarkerChannel.maybeAddPendingRequest(txnTopicPartition, metadata))
-      // TODO: Not sure this is the correct response here?
-      completionCallback(Errors.INVALID_TXN_STATE)
-    else {
-      val delayedTxnMarker = new DelayedTxnMarker(metadataToWrite, completionCallback)
-      txnMarkerPurgatory.tryCompleteElseWatch(delayedTxnMarker, Seq(metadata.pid))
-
-      val result = metadataToWrite.state match {
-        case PrepareCommit => TransactionResult.COMMIT
-        case PrepareAbort => TransactionResult.ABORT
-        case s => throw new IllegalStateException("Unexpected txn metadata state while writing markers: " + s)
+  def addTxnMarkersToSend(transactionalId: String,
+                          coordinatorEpoch: Int,
+                          txnResult: TransactionResult,
+                          txnMetadata: TransactionMetadata,
+                          newMetadata: TransactionMetadataTransition): Unit = {
+
+    def appendToLogCallback(error: Errors): Unit = {
+      error match {
+        case Errors.NONE =>
+          trace(s"Completed sending transaction markers for $transactionalId as $txnResult")
+
+          txnStateManager.getTransactionState(transactionalId) match {
+            case Some(epochAndMetadata) =>
+              epochAndMetadata synchronized {
```

Comment:
Actually I just realized we do not need to check the coordinator epoch anymore, since it is already checked in `appendTransactionToLog#updateCacheCallback`, which is executed before calling this callback, so the error code returned is already reflecting the fact if 

```
(epochAndMetadata.coordinatorEpoch == coordinatorEpoch && metadata.completeTransitionTo(newMetadata))
```

So we can directly go head with the error code here.